### DISCUSSION TITLE
layout with variable from the stash

### DISCUSSION THREAD

**Discussion by Haf-Linger** [mojolicious/mojo#1801] [Q&A] [answered]
I am trying to set the refresh rate of a status page using a stash variable
So I have the following code in the controller to render the page:
```
my $refreshrate = 10;
$self->stash( refreshrate => $refreshrate);
$self->render();
```
in the html template I have:
```
% layout 'refreshLayout', title => 'A title';
% content_for refresh => begin
	<meta http-equiv="refresh" content="<%= $refreshrate %>"/>
% end
<div>....body content .../<div>
```
and in the layout template I have:
```
<!DOCTYPE html>
<html>
  <head>
	<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=0"/>
       <meta charset="UTF-8"/>
	<%= content 'refresh' %>
        .....
 </head>
  <body><%= content %></body>
</html>
```

I always end up with a page rendered with no value in the content attribute of the refresh meta tag, but the meta tag will be there. So that when I replace in the template of the page the `"<%= $refreshrate %>`  with a fixed value like `'10'` then all is OK.

So what am I doing wrong here?
And while we are here,  why in the layout template do we do use <%= content %> (and not $content)
While in the template of the regular page we use the more logical <%= $variable %>
How does the replacement work in the layout template versus the replacement in the template of the main page?

**Comment by Haf-Linger** [comment]
Hmm, no answer so far. Maybe its me...?
Any suggestion on how to better phrase this question so that I do get an answer?

**Comment by rage311** [comment, marked as answer]
This seems to work as expected:

```
#!/usr/bin/env perl

use 5.034;
use Mojolicious::Lite -strict, -signatures;

get '/' => sub ($c) { $c->stash(refreshrate => 10) } => 'index';

app->start;

__DATA__

@@ index.html.ep
% layout 'refreshLayout', title => 'A title';
% content_for refresh => begin
  <meta http-equiv="refresh" content="<%= $refreshrate %>"/>
% end
<div>....body content ...</div>

@@ layouts/refreshLayout.html.ep
<!DOCTYPE html>
<html>
  <head>
    <meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=0"/>
    <meta charset="UTF-8"/>
    <%= content 'refresh' %>
  </head>
  <body><%= content %></body>
</html>

```

Generates the HTML content:

```
<!DOCTYPE html>
<html>
  <head>
	<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1, user-scalable=0"/>
       <meta charset="UTF-8"/>
		<meta http-equiv="refresh" content="10"/>

 </head>
  <body><div>....body content ...</div>
</body>
</html>
```

**Reply by Haf-Linger** [reply]
@rage311 
Thank you for your answer.
I can confirm that things work the way they should and I was able to find my real problem.

It seems to me that you can not use the same key names in the $app->defaults hash as in your $app->config hash.
I had the key `refreshrate` sitting in my `config` file in order to set the initial refreshrate.
At startup I moved this value from the `config` hash to the `defaults` hash (as I wanted it to be modifiable and I do not like to change the values that sit in the config hash).
So in reality my controller code looked like:
```
my $refreshrate = $self->defaults->{'refreshrate'};
$self->stash( refreshrate => $refreshrate);
$self->render();
```

turns out that if you do that, the `stash` will have a key `refreshrate` in the `config` part of the `stash`.
but there will be no key called `refreshrate` in the rest of the `stash`.
as soon as you change the name in the `defaults` hash to something different like `refresh`  things work OK.
So changing your setup and your controller code to:
```
my $refreshrate = $self->defaults->{'refresh'};
$self->stash( refreshrate => $refreshrate);
$self->render();
```

Am I dreaming?
And if not, is it behaviour intentional or a bug?
